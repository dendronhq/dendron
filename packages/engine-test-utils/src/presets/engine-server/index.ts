import { ENGINE_CONFIG_PRESETS } from "./config";
import { ENGINE_DELETE_PRESETS } from "./delete";
import { ENGINE_INFO_PRESETS } from "./info";
import { ENGINE_INIT_PRESETS } from "./init";
import { ENGINE_GET_NOTE_BLOCKS_PRESETS } from "./getNoteBlocks";
import NOTE_REF from "./note-refs";
import { ENGINE_QUERY_PRESETS } from "./query";
import { ENGINE_RENAME_PRESETS } from "./rename";
import { ENGINE_BULK_WRITE_NOTES_PRESETS } from "./bulkWriteNotes";
import { ENGINE_RENDER_PRESETS } from "./render";
import { ENGINE_WRITE_PRESETS, ENGINE_WRITE_PRESETS_MULTI } from "./write";
import _ from "lodash";
import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";

export { ENGINE_HOOKS, ENGINE_HOOKS_BASE, ENGINE_HOOKS_MULTI } from "./utils";
export { ENGINE_RENAME_PRESETS };
export { ENGINE_QUERY_PRESETS };
export { ENGINE_WRITE_PRESETS };
export { ENGINE_CONFIG_PRESETS };

export const ENGINE_SERVER = {
  NOTE_REF,
  ENGINE_WRITE_PRESETS,
  ENGINE_INIT_PRESETS,
  ENGINE_DELETE_PRESETS,
  ENGINE_INFO_PRESETS,
  ENGINE_RENAME_PRESETS,
  ENGINE_GET_NOTE_BLOCKS_PRESETS,
  ENGINE_QUERY_PRESETS,
  ENGINE_BULK_WRITE_NOTES_PRESETS,
  ENGINE_RENDER_PRESETS,
};

type TestPresetEntry = TestPresetEntryV4;
type TestPresetDict = { [key: string]: TestPresetEntry };

/**
 * 
 @example
 *  test("", async () => {
 *    const TestCase= getPreset({presets: ENGINE_PRESETS, nodeType, presetName: "init", key: 'BAD_SCHEMA'})
 *    const { testFunc, ...opts } = TestCase;;
 *    await runEngineTestV5(testFunc, { ...opts, expect });
 *});
 * @param param0 
 * @returns 
 */
export const getPreset = ({
  presets,
  presetName,
  nodeType,
  key,
}: {
  presets: typeof ENGINE_PRESETS;
  presetName: string;
  nodeType: "SCHEMAS" | "NOTES";
  key: string;
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  const out = _.get(ent.presets[nodeType], key);
  if (!out) {
    throw Error(`no key ${key} found in ${presetName}`);
  }
  return out;
};

export const getPresetMulti = ({
  presets,
  presetName,
  nodeType,
  key,
}: {
  presets: typeof ENGINE_PRESETS_MULTI;
  presetName: string;
  nodeType: "NOTES";
  key: string;
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  const out = _.get(ent.presets[nodeType], key);
  if (!out) {
    throw Error(`no key ${key} found in ${presetName}`);
  }
  return out;
};

export const getPresetGroup = ({
  presets,
  presetName,
  nodeType,
}: {
  presets: typeof ENGINE_PRESETS;
  presetName: string;
  nodeType: "SCHEMAS" | "NOTES";
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  return ent.presets[nodeType] as TestPresetDict;
};

// ^iygzn9r2758w
export const ENGINE_PRESETS = [
  { name: "write", presets: ENGINE_SERVER.ENGINE_WRITE_PRESETS },
  {
    name: "bulkWriteNotes",
    presets: ENGINE_SERVER.ENGINE_BULK_WRITE_NOTES_PRESETS,
  },
  { name: "delete", presets: ENGINE_SERVER.ENGINE_DELETE_PRESETS },
  { name: "rename", presets: ENGINE_SERVER.ENGINE_RENAME_PRESETS },
  { name: "init", presets: ENGINE_SERVER.ENGINE_INIT_PRESETS },
  { name: "query", presets: ENGINE_SERVER.ENGINE_QUERY_PRESETS },
  { name: "info", presets: ENGINE_SERVER.ENGINE_INFO_PRESETS },
  {
    name: "getNoteBlocks",
    presets: ENGINE_SERVER.ENGINE_GET_NOTE_BLOCKS_PRESETS,
  },
  { name: "render", presets: ENGINE_SERVER.ENGINE_RENDER_PRESETS },
];

export const ENGINE_PRESETS_MULTI = [
  { name: "write", presets: ENGINE_WRITE_PRESETS_MULTI },
  { name: "delete", presets: ENGINE_SERVER.ENGINE_DELETE_PRESETS },
  { name: "rename", presets: ENGINE_SERVER.ENGINE_RENAME_PRESETS },
  { name: "init", presets: ENGINE_SERVER.ENGINE_INIT_PRESETS },
  { name: "query", presets: ENGINE_SERVER.ENGINE_QUERY_PRESETS },
];
