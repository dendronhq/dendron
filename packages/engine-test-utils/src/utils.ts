import {
  CONSTANTS,
  DendronConfig,
  VaultUtils,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings,
} from "@dendronhq/common-all";
import { readYAML } from "@dendronhq/common-server";
import { AssertUtils } from "@dendronhq/common-test-utils";
import { DConfig, Git } from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";

export async function checkString(body: string, ...match: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      match,
    })
  ).toBeTruthy();
}

export async function checkNotInString(body: string, ...nomatch: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      nomatch,
    })
  ).toBeTruthy();
}

const getWorkspaceFolders = (wsRoot: string) => {
  const wsPath = path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME);
  const settings = fs.readJSONSync(wsPath) as WorkspaceSettings;
  return _.toArray(settings.folders);
};

export function checkVaults(opts: WorkspaceOpts, expect: any) {
  const { wsRoot, vaults } = opts;
  const configPath = DConfig.configPath(opts.wsRoot);
  const config = readYAML(configPath) as DendronConfig;
  expect(_.sortBy(config.vaults, ["fsPath", "workspace"])).toEqual(
    _.sortBy(vaults, ["fsPath", "workspace"])
  );
  const wsFolders = getWorkspaceFolders(wsRoot);
  expect(wsFolders).toEqual(
    vaults.map((ent) => {
      const out: WorkspaceFolderRaw = { path: VaultUtils.getRelPath(ent) };
      if (ent.name) {
        out.name = ent.name;
      }
      return out;
    })
  );
}

export class GitTestUtils {
  static async createRepoForWorkspace(wsRoot: string) {
    const git = new Git({ localUrl: wsRoot });
    await git.init();
    await git.add("dendron.yml");
    await git.commit({ msg: "init" });
  }

  /** Creates a "bare" git repository, to be used as the remote for a workspace.
   *
   * You'll probably want to just use `createRepoForRemoteWorkspace` instead, but this is provided if you are testing something it can't handle.
   */
  static async remoteCreate(remoteDir: string) {
    const git = new Git({ localUrl: remoteDir, bare: true });
    await git.init();
  }

  /** Adds a bare repository created with `createRemote` as the remote for the workspace.
   *
   * You'll probably want to just use `createRepoForRemoteWorkspace` instead, but this is provided if you are testing something it can't handle.
   */
  static async remoteAdd(wsRoot: string, remoteDir: string) {
    const git = new Git({ localUrl: wsRoot, remoteUrl: remoteDir });
    await git.remoteAdd();
    // Need to push to be able to set up remote tracking branch
    await git.push({ remote: "origin", branch: await git.getCurrentBranch() });
  }

  /** Set up a workspace with a remote, intended to be used when testing pull or push functionality.
   *
   * @param wsRoot Directory where the workspace will be stored.
   * @param remoteDir Directory where the remote will be stored. The workspace will pull and push to this remote.
   */
  static async createRepoForRemoteWorkspace(wsRoot: string, remoteDir: string) {
    await this.createRepoForWorkspace(wsRoot);
    await this.remoteCreate(remoteDir);
    await this.remoteAdd(wsRoot, remoteDir);
  }

  static async createRepoWithReadme(root: string) {
    const git = new Git({ localUrl: root });
    await git.init();
    const readmePath = path.join(root, "README.md");
    fs.ensureFileSync(readmePath);
    await git.add(".");
    await git.commit({ msg: "init" });
  }
}
