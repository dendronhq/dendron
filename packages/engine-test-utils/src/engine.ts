import {
  CleanDendronSiteConfig,
  CONSTANTS, DendronConfig, DEngineClient,
  DVault,
  DWorkspace,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings
} from "@dendronhq/common-all";
import {
  getDurationMilliseconds,
  tmpDir,
  vault2Path
} from "@dendronhq/common-server";
import {
  GenTestResults,
  PostSetupHookFunction,
  PreSetupHookFunction,
  RunEngineTestFunctionOpts,
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
  SetupTestFunctionV4,
  sinon,
  TestResult
} from "@dendronhq/common-test-utils";
import { LaunchEngineServerCommand } from "@dendronhq/dendron-cli";
import {
  createEngine as engineServerCreateEngine,
  DConfig,
  WorkspaceService
} from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import os from "os";
import path from "path";
import { SinonStub } from "sinon";
import { ENGINE_HOOKS } from "./presets";
import { GitTestUtils } from "./utils";

export type TestSetupWorkspaceOpts = {
  /**
   * Vaults to initialize engine with
   * Defaults to following if not set
   * [
   *    { fsPath: "vault1" },
   *    { fsPath: "vault2" },
   *    { fsPath: "vault3", name: "vaultThree" },
   *  ]
   */
  vaults?: DVault[];
  /**
   * Modify dendron config before initialization
   */
  modConfigCb?: (config: DendronConfig) => DendronConfig;
};

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<{engine: DEngineClient, port?: number}>;

/**
 * Create an {@link DendronEngine}
 */
export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return {engine: engineServerCreateEngine(opts) as DEngineClient, port: undefined};
}

export { DEngineClient, DVault, WorkspaceOpts };

/**
 * Create a server
 * @param opts
 * @returns
 */
// @ts-ignore
export async function createServer(opts: WorkspaceOpts) {
  return await new LaunchEngineServerCommand().enrichArgs({
    wsRoot: opts.wsRoot,
  });
}

/**
 * Create an {@link DendronEngineClient}
 */
export async function createEngineFromServer(
  opts: WorkspaceOpts
): Promise<any> {
  const { engine, port } = await createServer(opts);
  await engine.init();
  return {engine, port};
}

export function createSiteConfig(
  opts: Partial<CleanDendronSiteConfig> &
    Required<Pick<CleanDendronSiteConfig, "siteRootDir" | "siteHierarchies">>
): CleanDendronSiteConfig {
  let copts = {
    siteNotesDir: "docs",
    siteUrl: "https://localhost:8080",
    ...opts,
  };
  return {
    ...copts,
    siteIndex: DConfig.getSiteIndex(copts),
  };
}

/**
 *
 * @param opts.asRemote: add git repo
 * @param opts.wsRoot: override given wsRoot
 * @returns
 */
export async function setupWS(opts: {
  vaults: DVault[];
  workspaces?: DWorkspace[];
  asRemote?: boolean;
  wsRoot?: string;
}) {
  const wsRoot = opts.wsRoot || tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  ws.createConfig();
  const config = ws.config;
  let vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault, config, updateConfig: false });
      return vault;
    })
  );
  config.vaults = _.sortBy(config.vaults, "fsPath");
  if (config.site.duplicateNoteBehavior) {
    config.site.duplicateNoteBehavior.payload = (
      config.site.duplicateNoteBehavior.payload as string[]
    ).sort();
  }
  ws.setConfig(config);
  if (opts.workspaces) {
    const vaultsFromWs = await _.reduce(
      opts.workspaces,
      async (resp, ent) => {
        await resp;
        await WorkspaceService.createWorkspace({
          wsRoot: path.join(wsRoot, ent.name),
          vaults: ent.vaults,
        });
        return ws.addWorkspace({ workspace: ent });
      },
      Promise.resolve({ vaults: [] } as { vaults: DVault[] })
    );
    vaults = vaults.concat(vaultsFromWs.vaults);
  }
  if (opts.asRemote) {
    await GitTestUtils.createRepoWithReadme(wsRoot);
  }
  return { wsRoot, vaults };
}

export type RunEngineTestV5Opts = {
  preSetupHook?: SetupHookFunction;
  createEngine?: AsyncCreateEngineFunction;
  extra?: any;
  expect: any;
  workspaces?: DWorkspace[];
  setupOnly?: boolean;
  initGit?: boolean;
  initHooks?: boolean;
  addVSWorkspace?: boolean;
  /**
   * Path to preset wsRoot
   */
  wsRoot?: string;
} & TestSetupWorkspaceOpts;

export type RunEngineTestFunctionV5<T = any> = (
  opts: RunEngineTestFunctionOpts & { extra?: any; engineInitDuration: number, port?: number }
) => Promise<TestResult[] | void | T>;

export class TestPresetEntryV5 {
  public preSetupHook: PreSetupHookFunction;
  public postSetupHook: PostSetupHookFunction;
  public testFunc: RunEngineTestFunctionV4;
  public extraOpts: any;
  public setupTest?: SetupTestFunctionV4;
  public genTestResults?: GenTestResults;
  public vaults: DVault[];
  public workspaces: DWorkspace[];

  constructor(
    func: RunEngineTestFunctionV5,
    opts?: {
      preSetupHook?: PreSetupHookFunction;
      postSetupHook?: PostSetupHookFunction;
      extraOpts?: any;
      setupTest?: SetupTestFunctionV4;
      genTestResults?: GenTestResults;
      vaults?: DVault[];
    }
  ) {
    let {
      preSetupHook,
      postSetupHook,
      extraOpts,
      setupTest,
      genTestResults,
      workspaces,
    } = _.defaults(opts, {
      workspaces: [],
    });
    this.preSetupHook = preSetupHook ? preSetupHook : async () => {};
    this.postSetupHook = postSetupHook ? postSetupHook : async () => {};
    this.testFunc = _.bind(func, this);
    this.extraOpts = extraOpts;
    this.setupTest = setupTest;
    this.genTestResults = _.bind(
      genTestResults ? genTestResults : async () => [],
      this
    );
    this.workspaces = workspaces;
    this.vaults = opts?.vaults || [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      {
        name: "vaultThree",
        fsPath: "vault3",
      },
    ];
  }
}

/**
 *
 * To create empty workspace, initilizae with `vaults = []`
 * @param func
 * @param opts.vaults: By default, initiate 3 vaults {vault1, vault2, (vault3, "vaultThree")}
 * @param opts.preSetupHook: By default, initiate empty
 * @param opts.wsRoot: Override the randomly generated test directory for the wsRoot
 * @returns
 */
export async function runEngineTestV5(
  func: RunEngineTestFunctionV5,
  opts: RunEngineTestV5Opts
): Promise<any> {
  const {
    preSetupHook,
    extra,
    vaults: vaultsInit,
    createEngine,
    initGit,
    workspaces,
    addVSWorkspace,
  } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    // third vault has diff name
    vaults: [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      { fsPath: "vault3", name: "vaultThree" },
    ],
    addVSWorkspace: false,
  });
  try {
    // make sure tests don't overwrite local homedir contents
    TestEngineUtils.mockHomeDir();
    const { wsRoot, vaults } = await setupWS({
      vaults: vaultsInit,
      workspaces,
      wsRoot: opts.wsRoot,
    });
    if ((opts.initHooks, vaults)) {
      fs.ensureDirSync(path.join(wsRoot, CONSTANTS.DENDRON_HOOKS_BASE));
    }
    await preSetupHook({ wsRoot, vaults });
    const resp = await createEngine({ wsRoot, vaults });
    const engine = resp.engine;
    const start = process.hrtime();
    const initResp = await engine.init();
    if (addVSWorkspace) {
      fs.writeJSONSync(
        path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME),
        {
          folders: vaults.map((ent) => ({
            path: ent.fsPath,
            name: ent.name,
          })) as WorkspaceFolderRaw[],
          settings: {},
          extensions: {},
        } as WorkspaceSettings,
        { spaces: 4 }
      );
    }
    const engineInitDuration = getDurationMilliseconds(start);
    const testOpts = {
      wsRoot,
      vaults,
      engine,
      initResp,
      port: resp.port,
      extra,
      config: engine,
      engineInitDuration,
    };
    if (initGit) {
      await GitTestUtils.createRepoForWorkspace(wsRoot);
      await Promise.all(
        vaults.map((vault) => {
          return GitTestUtils.createRepoWithReadme(
            vault2Path({ vault, wsRoot })
          );
        })
      );
    }
    if (opts.setupOnly) {
      return testOpts;
    }
    const results = (await func(testOpts)) || [];
    await runJestHarnessV2(results, expect);
    return { opts: testOpts, resp: undefined, wsRoot };
  } finally {
    // restore sinon so other tests can keep running
    sinon.restore();
  }
}

export function testWithEngine(
  prompt: string,
  func: RunEngineTestFunctionV4,
  opts?: Omit<RunEngineTestV5Opts, "expect"> & { only?: boolean }
) {
  if (opts?.only) {
    return test.only(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  } else {
    return test(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  }
}

export class TestEngineUtils {
  static mockHomeDir(dir?: string): SinonStub {
    if (_.isUndefined(dir)) dir = tmpDir().name;
    return sinon.stub(os, "homedir").returns(dir);
  }

  static vault1(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault1" })!;
  }

  static vault2(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault2" })!;
  }

  static vault3(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault3" })!;
  }
}
