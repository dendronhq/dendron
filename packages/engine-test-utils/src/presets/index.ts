import { DVault, VaultUtils } from "@dendronhq/common-all";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { WorkspaceOpts } from "../engine";

export * from "./engine-server";
export * from "./pods-core";
export * from "./vaults";

export enum SETUP_HOOK_KEYS {
  /**
   * alpha: link(beta)
   * beta: link(alpha)
   */
  WITH_LINKS = "WITH_LINKS",
}

const createLink = (fname: string, opts: { vaultPrefix?: DVault }) => {
  let cVaultPrefix = "";
  if (opts.vaultPrefix) {
    cVaultPrefix = VaultUtils.toURIPrefix(opts.vaultPrefix) + "/";
  }
  return `[[${cVaultPrefix}${fname}]]`;
};

export async function callSetupHook(
  key: SETUP_HOOK_KEYS,
  opts: {
    workspaceType: "single" | "multi";
    withVaultPrefix?: boolean;
  } & WorkspaceOpts
) {
  const { workspaceType, vaults, wsRoot, withVaultPrefix } = opts;
  const isMultiVault = workspaceType !== "single";
  const cVaults = isMultiVault ? vaults : [vaults[0], vaults[0]];

  // WITH LINKS
  if (key === SETUP_HOOK_KEYS.WITH_LINKS) {
    const link1 = createLink("beta", {
      vaultPrefix: withVaultPrefix ? cVaults[1] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "alpha",
      wsRoot,
      vault: cVaults[0],
      body: link1,
    });
    const link2 = createLink("alpha", {
      vaultPrefix: withVaultPrefix ? cVaults[0] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "beta",
      wsRoot,
      vault: cVaults[1],
      body: link2,
    });
    if (isMultiVault) {
      await NoteTestUtilsV4.createNote({
        fname: "beta",
        wsRoot,
        vault: cVaults[0],
        body: link2,
        genRandomId: true,
      });
    }
  } else {
    throw Error("not supported key");
  }
}
