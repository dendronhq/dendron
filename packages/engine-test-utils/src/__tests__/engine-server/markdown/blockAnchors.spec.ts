import { ConfigService, ConfigUtils, URI } from "@dendronhq/common-all";
import { AssertUtils, TestPresetEntryV4 } from "@dendronhq/common-test-utils";
import { Parent, Text } from "@dendronhq/engine-server";
import {
  BlockAnchor,
  DendronASTDest,
  DendronASTTypes,
  MDUtilsV5,
  UnistNode,
} from "@dendronhq/unified";
import _ from "lodash";
import { TestConfigUtils } from "../../..";
import { runEngineTestV5 } from "../../../engine";
import { ENGINE_HOOKS } from "../../../presets";
import { TestUnifiedUtils } from "../../../utils";
import { createProcForTest, createProcTests, ProcTests } from "./utils";

const { getDescendantNode } = TestUnifiedUtils;

function proc() {
  return MDUtilsV5.procRemarkParseNoData({}, { dest: DendronASTDest.HTML });
}

function runAllTests(opts: { name: string; testCases: ProcTests[] }) {
  const { name, testCases } = opts;
  describe(name, () => {
    test.each(
      testCases.map((ent) => [`${ent.dest}: ${ent.name}`, ent.testCase])
      // @ts-ignore
    )("%p", async (_key, testCase: TestPresetEntryV4) => {
      await runEngineTestV5(testCase.testFunc, {
        expect,
        preSetupHook: async (opts) => {
          await testCase.preSetupHook(opts);
          await TestConfigUtils.withConfig(
            (config) => {
              const defaultConfig = ConfigUtils.genDefaultConfig();
              ConfigUtils.setVaults(
                defaultConfig,
                ConfigUtils.getVaults(config)
              );
              return defaultConfig;
            },
            { wsRoot: opts.wsRoot }
          );
        },
      });
    });
  });
}

function getBlockAnchor(node: UnistNode): BlockAnchor {
  return getDescendantNode<BlockAnchor>(expect, node, 0, 0);
}

describe("blockAnchors", () => {
  describe("parse", () => {
    test("parses anchor by itself", () => {
      const resp = proc().parse("^block-0-id");
      expect(getBlockAnchor(resp).type).toEqual(DendronASTTypes.BLOCK_ANCHOR);
      expect(getBlockAnchor(resp).id).toEqual("block-0-id");
    });

    test("doesn't parse inline code block", () => {
      const resp = proc().parse("`^block-id`");
      expect(getBlockAnchor(resp).type).toEqual("inlineCode");
    });

    test("doesn't parse block anchor inside a link", async () => {
      const resp = proc().parse("[[#^block-id]]");
      expect(getBlockAnchor(resp).type).toEqual(DendronASTTypes.WIKI_LINK);
    });

    test("parses a block anchor in the middle of a paragraph", async () => {
      const resp = proc().parse(
        ["Lorem ipsum ^block-id", "dolor amet."].join("\n")
      );
      expect(getDescendantNode(expect, resp, 0, 1).type).toEqual(
        DendronASTTypes.BLOCK_ANCHOR
      );
    });

    test("doesn't parse code block not at the end of the line", async () => {
      const resp = proc().parse("^block-id Lorem ipsum");
      expect(getDescendantNode(expect, resp, 0, 0).type).toEqual("text");
      expect(
        getDescendantNode<Parent>(expect, resp, 0).children.length
      ).toEqual(1); // text only, nothing else
    });

    test("parses anchors at the end of the line", () => {
      const resp = proc().parse("Lorem ipsum ^block-id");
      const text = getDescendantNode(expect, resp, 0, 0);
      expect(text.type).toEqual("text");
      const anchor = getDescendantNode<BlockAnchor>(expect, resp, 0, 1);
      expect(anchor.type).toEqual(DendronASTTypes.BLOCK_ANCHOR);
      expect(anchor.id).toEqual("block-id");
    });

    test("parses anchors at the end of headers", () => {
      const resp = proc().parse("# Lorem ipsum ^block-id");
      const header = getDescendantNode<Text>(expect, resp, 0, 0);
      expect(header.value.trim()).toEqual("Lorem ipsum");
      const anchor = getDescendantNode<BlockAnchor>(expect, resp, 0, 1);
      expect(anchor.type).toEqual(DendronASTTypes.BLOCK_ANCHOR);
      expect(anchor.id).toEqual("block-id");
    });

    test("doesn't parse footnote as block anchor", async () => {
      const resp = proc().parse("Lorem ipsum [^footnote]");
      expect(getDescendantNode(expect, resp, 0, 1).type).toEqual(
        "footnoteReference"
      );
    });
  });

  describe("rendering", () => {
    const anchor = "^my-block-anchor-0";
    const SIMPLE = createProcTests({
      name: "simple",
      setupFunc: async ({ engine, vaults, extra, wsRoot }) => {
        const config = (
          await ConfigService.instance().readConfig(URI.file(wsRoot))
        )._unsafeUnwrap();
        const proc2 = await createProcForTest({
          engine,
          dest: extra.dest,
          vault: vaults[0],
          config,
        });
        const resp = await proc2.process(anchor);
        return { resp };
      },
      verifyFuncDict: {
        [DendronASTDest.HTML]: async ({ extra }) => {
          const { resp } = extra;
          expect(resp).toMatchSnapshot();
          return [
            {
              actual: await AssertUtils.assertInString({
                body: resp.toString(),
                match: ["<a", `href="#${anchor}"`, `id="${anchor}"`, "</a>"],
                nomatch: ["visibility: hidden"],
              }),
              expected: true,
            },
          ];
        },
        [DendronASTDest.MD_DENDRON]: async ({ extra }) => {
          const { resp } = extra;
          expect(resp).toMatchSnapshot();
          return [
            {
              actual: await AssertUtils.assertInString({
                body: resp.toString(),
                match: [anchor],
              }),
              expected: true,
            },
          ];
        },
      },
      preSetupHook: ENGINE_HOOKS.setupBasic,
    });

    const END_OF_PARAGRAPH = createProcTests({
      name: "end of paragraph",
      setupFunc: async ({ engine, vaults, extra, wsRoot }) => {
        const config = (
          await ConfigService.instance().readConfig(URI.file(wsRoot))
        )._unsafeUnwrap();
        const proc2 = await createProcForTest({
          engine,
          dest: extra.dest,
          vault: vaults[0],
          config,
        });
        const resp = await proc2.process(`Lorem ipsum dolor amet ${anchor}`);
        return { resp };
      },
      verifyFuncDict: {
        [DendronASTDest.HTML]: async ({ extra }) => {
          const { resp } = extra;
          expect(resp).toMatchSnapshot();
          return [
            {
              actual: await AssertUtils.assertInString({
                body: resp.toString(),
                match: ["<a", `href="#${anchor}"`, `id="${anchor}"`, "</a>"],
                nomatch: ["visibility: hidden"],
              }),
              expected: true,
            },
          ];
        },
      },
      preSetupHook: ENGINE_HOOKS.setupBasic,
    });

    const AFTER_CODE_BLOCK = createProcTests({
      name: "after code block",
      setupFunc: async ({ engine, vaults, extra, wsRoot }) => {
        const config = (
          await ConfigService.instance().readConfig(URI.file(wsRoot))
        )._unsafeUnwrap();
        const proc2 = await createProcForTest({
          engine,
          dest: extra.dest,
          vault: vaults[0],
          config,
        });
        const resp = await proc2.process(
          ["```", "const x = 1;", "```", "", anchor].join("\n")
        );
        return { resp };
      },
      verifyFuncDict: {
        [DendronASTDest.HTML]: async ({ extra }) => {
          const { resp } = extra;
          expect(resp).toMatchSnapshot();
          return [
            {
              actual: await AssertUtils.assertInString({
                body: resp.toString(),
                match: ["<a", `href="#${anchor}"`, `id="${anchor}"`, "</a>"],
                nomatch: ["visibility: hidden"],
              }),
              expected: true,
            },
          ];
        },
      },
      preSetupHook: ENGINE_HOOKS.setupBasic,
    });

    const AFTER_TABLE = createProcTests({
      name: "after table",
      setupFunc: async ({ engine, vaults, extra, wsRoot }) => {
        const config = (
          await ConfigService.instance().readConfig(URI.file(wsRoot))
        )._unsafeUnwrap();
        const proc2 = await createProcForTest({
          engine,
          dest: extra.dest,
          vault: vaults[0],
          config,
        });

        const p = proc2.parse(
          [
            "| t   | a   |",
            "| --- | --- |",
            "| c   | d   |",
            "",
            `${anchor}`,
          ].join("\n")
        );
        const n = await proc2.run(p);
        const resp = proc2.stringify(n);
        return { resp };
      },
      verifyFuncDict: {
        [DendronASTDest.HTML]: async ({ extra }) => {
          const { resp } = extra;
          expect(resp).toMatchSnapshot();
          return [
            {
              actual: await AssertUtils.assertInString({
                body: resp.toString(),
                match: ["<a", `href="#${anchor}"`, `id="${anchor}"`, "</a>"],
                nomatch: ["visibility: hidden"],
              }),
              expected: true,
            },
          ];
        },
      },
      preSetupHook: ENGINE_HOOKS.setupBasic,
    });

    const ALL_TEST_CASES = [
      ...SIMPLE,
      ...END_OF_PARAGRAPH,
      ...AFTER_CODE_BLOCK,
      ...AFTER_TABLE,
    ];
    runAllTests({ name: "compile", testCases: ALL_TEST_CASES });
  });
});
