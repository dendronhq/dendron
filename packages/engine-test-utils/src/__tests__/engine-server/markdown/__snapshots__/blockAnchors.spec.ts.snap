// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blockAnchors rendering compile "HTML: after code block" 1`] = `
VFile {
  "contents": "<h1 id=\\"root\\">Root</h1>
<p><a aria-hidden=\\"true\\" class=\\"block-anchor anchor-heading icon-link\\" id=\\"^my-block-anchor-0\\" href=\\"#^my-block-anchor-0\\"></a></p><pre><code>const x = 1;
</code></pre><p></p>
<p></p>
<hr>
<strong>Children</strong>
<ol>
<li><a href=\\"bar\\">Bar</a></li>
<li><a href=\\"foo\\">Foo</a></li>
</ol>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`blockAnchors rendering compile "HTML: after table" 1`] = `
"<h1 id=\\"root\\">Root</h1>
<div class=\\"table-responsive\\">












<table><thead><tr><th><a aria-hidden=\\"true\\" class=\\"block-anchor anchor-heading icon-link\\" id=\\"^my-block-anchor-0\\" href=\\"#^my-block-anchor-0\\"></a>t</th><th>a</th></tr></thead><tbody><tr><td>c</td><td>d</td></tr></tbody></table></div>
<p></p>
<hr>
<strong>Children</strong>
<ol>
<li><a href=\\"bar\\">Bar</a></li>
<li><a href=\\"foo\\">Foo</a></li>
</ol>"
`;

exports[`blockAnchors rendering compile "HTML: end of paragraph" 1`] = `
VFile {
  "contents": "<h1 id=\\"root\\">Root</h1>
<p><a aria-hidden=\\"true\\" class=\\"block-anchor anchor-heading icon-link\\" id=\\"^my-block-anchor-0\\" href=\\"#^my-block-anchor-0\\"></a>Lorem ipsum dolor amet </p>
<hr>
<strong>Children</strong>
<ol>
<li><a href=\\"bar\\">Bar</a></li>
<li><a href=\\"foo\\">Foo</a></li>
</ol>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`blockAnchors rendering compile "HTML: simple" 1`] = `
VFile {
  "contents": "<h1 id=\\"root\\"><a aria-hidden=\\"true\\" class=\\"block-anchor anchor-heading icon-link\\" id=\\"^my-block-anchor-0\\" href=\\"#^my-block-anchor-0\\"></a>Root</h1>
<p></p>
<hr>
<strong>Children</strong>
<ol>
<li><a href=\\"bar\\">Bar</a></li>
<li><a href=\\"foo\\">Foo</a></li>
</ol>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`blockAnchors rendering compile "MD_DENDRON: simple" 1`] = `
VFile {
  "contents": "# Root

^my-block-anchor-0

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;
